### PR TITLE
Implement profile specific settings for backup folder.

### PR DESCRIPTION
**Pull request by romoloman**
Profile backup path if set overrides general one, in any case the backup will be performed if either profile specific setting or general settings ones are enabled.
Disabling backup in profile will not disable backups if generic settings are enabled

### CODE REVIEW

File: companion/src/apppreferencesdialog.cpp [hunk 1]
```diff
@@ -164,6 +166,19 @@ void AppPreferencesDialog::initSettings()
   ui->stickmodeCB->setCurrentIndex(g.profile[g.id()].defaultMode());
   ui->renameFirmware->setChecked(g.profile[g.id()].renameFwFiles());
   ui->sdPath->setText(g.profile[g.id()].sdPath());
+  ui->profilebackupPath->setText(g.profile[g.id()].pBackupDir());
+  if (!g.profile[g.id()].pBackupDir().isEmpty()) {
+    if (QDir(g.profile[g.id()].pBackupDir()).exists()) {
+      ui->profilebackupPath->setText(g.backupDir());
```
Comment (projectkk2glider) [inline companion/src/apppreferencesdialog.cpp]: Here you overwrite profile backup folder with the application global backup folder. Why?

File: companion/src/mainwindow.cpp [hunk 2]
```diff
@@ -1336,16 +1352,36 @@ void MainWindow::writeFlash(QString fileToFlash)
...
+        else {
+          backupEnable = false;
         }
       }
+      if ((g.enableBackup() || g.profile[g.id()].penableBackup()) && !(backupEnable) && !IS_TARANIS(GetEepromInterface()->getBoard())) {
+        if (!QDir(backupPath).exists()) {
+          QMessageBox::warning(this, tr("Backup is impossible"), tr("The backup dir set in preferences does not exist"));
+        }      
+      } 
       else {
```
Comment (projectkk2glider) [inline companion/src/mainwindow.cpp]: Can we put this section and one in MainWindow::writeBackup() in a new function? They look very similar.

File: companion/src/mainwindow.cpp [hunk 3]
```diff
@@ -1379,13 +1415,35 @@ void MainWindow::writeFlash(QString fileToFlash)
...
+        }
+        else {
+          backupEnable = false;
+        }
+      }
+      if ((g.enableBackup() || g.profile[g.id()].penableBackup()) && !(backupEnable)) {
+        if (!QDir(backupPath).exists()) {
+          QMessageBox::warning(this, tr("Backup is impossible"), tr("The backup dir set in preferences does not exist"));
+        }      
       }
```
Comment (projectkk2glider) [inline companion/src/mainwindow.cpp]: Same as previous comment, we have three same sections of code.

File: companion/src/mdichild.cpp [hunk 4]
```diff
@@ -516,18 +516,35 @@ QString MdiChild::strippedName(const QString &fullFileName)
...
+    }
+    else {
+      backupEnable = false;
+    }
+  }
+  if ((g.enableBackup() || g.profile[g.id()].penableBackup()) && !(backupEnable)) {
+    if (!QDir(backupPath).exists()) {
+      QMessageBox::warning(this, tr("Backup is impossible"), tr("The backup dir set in preferences does not exist"));
+    }      
   }
```
Comment (projectkk2glider) [inline companion/src/mdichild.cpp]: Four :)

**Comment by projectkk2glider** [comment]
Just to cross-reference: this pull fixes #2080 on master branch

File: companion/src/apppreferencesdialog.cpp (hunk 1 above)
Comment (romoloman) [inline companion/src/apppreferencesdialog.cpp]: This is a mistake, thanks for reviewing... the code was already duplicated across files, i only added some more.
Truly we can set a function to get the proper backup path.

**Comment by bsongis** [comment]
@projectkk2glider May I let you press the button?

**Comment by romoloman** [comment]
I think that as in 2.1 things changes a lot makes little sense to move out of mainwindows.
As is mostly called there.

**Pull request merged**